sdc/native/hashmap.hpp

```diff
@@ -0,0 +1,907 @@
+            this->val_info.incref(val);
+        }
+        else
+        {
+            // insertion failed key already exists
+            this->val_info.decref(existing_node->second);
+            free(existing_node->second);
+            existing_node->second = _val;
+            this->val_info.incref(val);
+        }
+    }
+}
+
+template<typename Key,
+         typename HashCompare
+>
+void GenericHashmapType<Key, void*, HashCompare>::clear()
+{
+    for (auto kv_pair: (*this->p_map)) {
+        this->val_info.decref(kv_pair.second);
+        free(kv_pair.second);
+    }
+    this->p_map->clear();
+}
+
+
+template<typename Key,
+         typename HashCompare
+>
+int8_t GenericHashmapType<Key, void*, HashCompare>::pop(Key key, void* res) {
+    bool found = false;
+    {
+        typename map_type::const_accessor result;
+        found = this->p_map->find(result, key);
+        if (found)
+        {
+            memcpy(res, result->second, this->val_info.size);
+            free(result->second);
+            // no decref for value since it would be returned (and no incref on python side!)
+            this->p_map->erase(result);
+        }
+        result.release();
+    }
+
+    return found;
+}
+
+
+template<typename Val>
+void GenericHashmapType<void*, Val, VoidPtrHashCompare>::set(void* key, Val val)
+{
+    auto ksize = this->key_info.size;
+    void* _key = malloc(ksize);
+    memcpy(_key, key, ksize);
+
+    typename map_type::value_type inserted_node(_key, val);
+    {
+        typename map_type::accessor existing_node;
+        bool ok = this->p_map->insert(existing_node, inserted_node);
+        if (ok)
+        {
+            // isnertion succeeded need to incref key
+            this->key_info.incref(key);
+        }
+        else
+        {
+            // insertion failed key already exists
+            free(_key);
+            existing_node->second = val;
+        }
+    }
+}
+
+template<typename Val>
+void GenericHashmapType<void*, Val, VoidPtrHashCompare>::clear()
+{
+    for (auto kv_pair: (*this->p_map)) {
+        this->key_info.decref(kv_pair.first);
+        free(kv_pair.first);
+    }
+    this->p_map->clear();
+}
+
+template<typename Val>
+int8_t GenericHashmapType<void*, Val, VoidPtrHashCompare>::pop(void* key, void* res) {
+    bool found = false;
+    {
+        typename map_type::const_accessor result;
+        found = this->p_map->find(result, key);
+        if (found)
+        {
+            memcpy(res, &(result->second), this->val_info.size);
+            this->key_info.decref(result->first);
+            free(result->first);
+            // no decref for value since it would be returned (and no incref on python side!)
+            this->p_map->erase(result);
+        }
+        result.release();
+    }
+
+    return found;
+}
+
+
+void GenericHashmapType<void*, void*, VoidPtrHashCompare>::set(void* key, void* val)
+
+{
+    auto ksize = this->key_info.size;
+    void* _key = malloc(ksize);
```

Comment:
Why not new? Can we wrap these explicit mallocs/frees into `unique_ptr`?

Reply:
It's problematic. I've tried a bit, but using `unique_ptr<char[ksize]>` is not possible since ksize is not compile-time const. Also the fact that TBB hashmap has to store the raw pointer (since HashCompare is actually numba functions that expect key/value to be pointers to Numba structs) kind of decrease the value of it (i.e. the ownership of allocated memory should go to hashmap itself and this function frees this memory only if insertion failed).

Reply:
You can use unique_ptr with dynamically-sized arrays. Like this :
```C++
auto _key = std::unique_ptr<char[]>(new char[ksize])
```
Or like this:
```C++
auto _key = std::make_unique<char[]>(ksize); 
```